### PR TITLE
Query predicates Implementation

### PR DESCRIPTION
**Pull request by stackmystack**
I took the implementation from the Rust bindings.

I tried to preserve as much as possible the original structure/variables/etc to make comparison with the existing code easier.

I didn't find tests on predicates, but I will look around a second time.

What I found disturbing is the `any` family of predicates. I put the tests that pass, but I think there's something missing, the results don't make sense.

**TODO** (and will not be part of this MR):

1. Property Predicates:
    1. `set!`
    1. `is?` / `is-of?`

### CODE REVIEW

File: lib/tree_stand/parser.rb [hunk 1]
```diff
@@ -180,7 +180,7 @@ def self.language(name)
 
       # We know that the bindings will accept `lib`, but I don't know how to tell sorbet
       # the types in ext/tree_sitter where `load` is defined.
-      TreeSitter::Language.load(name.gsub(/-/, '_'), T.unsafe(lib))
+      TreeSitter::Language.load(name.gsub('-', '_'), T.unsafe(lib))
```
Comment (akimd) [inline lib/tree_stand/parser.rb:183]: I'm surprised RuboCop doesn't recommend `tr` here.

**Review by akimd — COMMENTED** [review]
I can't claim I grokked it all :)

File: lib/tree_sitter/helpers.rb [hunk 2]
```diff
@@ -0,0 +1,23 @@
+# frozen_string_literal: true
+
+# splits an array like [rust](https://doc.rust-lang.org/std/primitive.slice.html#method.split)
+def array_split_like_rust(array, &block)
+  return enum_for(__method__, array) if !block_given?
+
+  return [] if array.empty?
+
```
[the remark this answers is not in the document]
Comment (stackmystack) [reply lib/tree_sitter/helpers.rb:8]: It's not, because if you have `[1,2,3,4,5]` and you split on `3`, you get `[[1,2],[],[4,5]]`.

**Comment by stackmystack** [comment]
> I can't claim I grokked it all :)

Me neither xD

**Comment by stackmystack** [comment]
So I tested those predicates with the rust binding, and I can't say they're intuitive. It looks like this implementation is iso.


### `Cargo.toml`

```toml
# …
[dependencies]
tree-sitter = "0.22"
tree-sitter-rust = "0.21"

```

### `src/main.src`

```rust
use tree_sitter::{Parser, Query, QueryCursor};

fn main() {
    let mut parser = Parser::new();
    let language = tree_sitter_rust::language();
    parser
        .set_language(&language)
        .expect("Error loading Rust grammar");
    let source_code = "[1,1,1,1]";
    let mut tree = parser.parse(source_code, None).unwrap();
    let root_node = tree.root_node();

    println!("{:?}", root_node.to_sexp());

    let query_str = r#"((integer_literal)* @b (#any-not-eq? @b "1"))"#;

    let query = Query::new(&language, query_str).unwrap();
    let mut cursor = QueryCursor::new();
    let matches = cursor.matches(&query, root_node, source_code.as_bytes());
    let res: Vec<_> = matches.collect();
    println!("{}", res.len());

    let mut idx = 0;
    for m in res {
        println!("{}:", idx);

        for c in m.captures {
            let b = c.node.byte_range();
            let s: Vec<_> = source_code
                .bytes()
                .skip(b.start)
                .take(b.end - b.start)
                .collect();
            let txt = String::from_utf8(s).unwrap();
            println!("  {}:", query.capture_names()[c.index as usize]);
            println!(
                "      {} | {} [{}, {}]",
                c.node.to_sexp(),
                txt,
                c.node.start_byte(),
                c.node.end_byte()
            );
        }

        idx += 1;
    }
}
```

I will revisit later to test better.

**Comment by stackmystack** [comment]
@DerekStride I executed `bin/tapioca todo` to correct something sorbet is not seeing (`QueryPredicate::*`).

I got a message saying it's deprecated and I should use `bin/tapioca gems`, but this generates a lot of files, and I am not sure how to approach sorbet and work with it …

You have any recommendations?

**Comment by DerekStride** [comment]
I usually only commit the generated files that are needed but that's quite tedious. There's a few options that might work to simplify this process.

1. Commit all generated files
2. Add the generated files to the gitignore and generate them in the setup script
3. Use `todo` for now and we can migrate to using `gems` later

**Comment by stackmystack** [comment]
Thanks @DerekStride, I just added the todo for now …

**Pull request merged**